### ISSUE TITLE
Using wm8731 instead of PCM5012a

### ISSUE THREAD

**Issue by okyeron** [chrislatina/MachineListening#7]
I'm building a Terminal Tedium this week and the most recent version uses the wm8731 DAC instead of the PCM5012a.

Will the wm8731 DAC be compatible with this project, or will there need to be some modifications to the code?

**Comment by chrislatina** [comment]
That's a good question -- I'm working on some of the new revisions myself but haven't actually completed any quite yet.

**Comment by okyeron** [comment]
OK. I'll report back when I get my build up and running.

**Comment by okyeron** [comment]
OK - I've got the TT up and running with the wm8731. This DAC gives the TT audio inputs, so I *think* all the setup for the USB sound card should be moot now.

However, there may be some SamplingRate issues. The wm8731 is at 48000 and I get the following errors trying to run mycc

`Device Info: snd_rpi_proto: - (hw:0,0), Host API: 0, SampleRate: 48000.000000
Device Info: sysdefault, Host API: 0, SampleRate: 48000.000000
Device Info: dmix, Host API: 0, SampleRate: 48000.000000
Device Info: default, Host API: 0, SampleRate: 48000.000000
Expression 'paInvalidSampleRate' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2043
Expression 'PaAlsaStreamComponent_InitialConfigure( &self->playback, outParams, self->primeBuffers, hwParamsPlayback, &realSr )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2717
Expression 'PaAlsaStream_Configure( stream, inputParameters, outputParameters, sampleRate, framesPerBuffer, &inputLatency, &outputLatency, &hostBufferSizeMode )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2838
An error occured while using the portaudio stream
Error number: -9997
Error message: Invalid sample rate
`

**Comment by okyeron** [comment]
Updated main.cpp with 48000 SAMPLE_RATE and changed the inputParameters.device = 0 and I've got something working now. Audio goes thru and is processed in some fashion, but I'm not entirely sure what should be the end result just yet.

**Comment by chrislatina** [comment]
Great -- So you should start seeing triggers generated when running audio into the TT. I assume you did end up setting up PortAudio -- it is required. There is also a simple synthesis library in which features can be routed to white noise, sine waves, etc.

**Comment by okyeron** [comment]
Yeah, I was unclear if PortAudio was working or not and if the audio inputs were just the default device, or something else. Took me awhile to sort that out.

Output - I'm getting *something* generated when I run audio in, but should I expect the output to be "audio" or some trigger voltage I should patch somewhere else? (not sure what you mean by "seeing triggers") 

A "hello world" or quick start example patch description would be helpful to know if it's working as expected.

**Comment by chrislatina** [comment]
So the Digital outputs will send CV gates / triggers. I'm not sure if the GPIO pin config is the same between revisions but I expect so. My apologies not in front of this device or code base at the moment.

https://github.com/mxmxmx/terminal_tedium/wiki/hardware-basics

**Comment by okyeron** [comment]
No worries. Thanks for the responses. I think the pin config is the same - I suppose I'd need to check your code to see where the GPIO pins are assigned?.

**Comment by okyeron** [comment]
If I wanted to flip the Audio Output and the Analog Output jacks (L/R Audio outputs) - can I reassign those in the code? Can't seem to find it.

(your orientation is flipped from the current TT panel designs)

**Comment by chrislatina** [comment]
The outputs are hard-wired from specific GPIO pins to the DAC to the jack itself. Those mapping must be static but obviously correctly mapped -- again I'm not sure if this has changed with the new rev. IE, Digital Output is direct 0/1/PWM from the GPIO pin while the Audio Output runs PCM to the DAC chip to the L/R Audio jacks. Do you just mean flipping Left with Right!? Go ahead! I'm completely open to variation on the orientation, etc. I actually used mine "upside-down" because of the way the pots were oriented.

**Comment by okyeron** [comment]
I just meant swapping L/R.  I managed to hack this by swapping the order of the *out++ calls to portaudio.

Additional question - My "clean" audio output is not clean at all. It's very distorted or something. I can't figure out why. Output from pd patches or Alsa are clean audio so it seems to be something in the Listening code or portaudio. 

Is it possible that because the input was changed to stereo, that the stream is getting doubled or something? 

There's a comment in the code about  "add ++ for stereo" but I don't understand what to do in this case. (I'm still a newb at reading this kind of code).

**Comment by okyeron** [comment]
Got my audio cleaned up, so all is good. Might still need some tweaking tho. Will update my version soon.
